System/Console/Haskeline.hs

```diff
@@ -244,8 +245,7 @@ getPassword x = promptedInput
                                         $ runCommandLoop tops prefix loop
                                         $ Password [] x)
                     (\fops -> let h_in = inputHandle fops
-                              in bracketSet (hGetEcho h_in) (hSetEcho h_in) False
-                                  $ runMaybeT $ getLocaleLine fops)
+                              in withoutInputEcho h_in $ runMaybeT $ getLocaleLine fops)
```

Comment:
Ah, got it.  What I meant is that the inputHandle would still be there implicitly, but be encapsulated within the closure.  (Which is already happening for other fields in `FileOps`, e.g., `getLocaleLine`.)

In `System.Console.Haskeline.Term`:

    data RunTerm { ..., termOps = Either TermOps FileOps }
    data FileOps = FileOps {
            withoutInputEcho :: forall a . IO a -> IO a,  -- This is the new line
            wrapFileInput :: forall a . IO a -> IO a,  -- This is an existing field
            ...
        }

and in `System.Console.Haskeline.Backend.Win32`:
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend/Win32.hsc#L403

    fileRunTerm :: Handle -> IO RunTerm
    fileRunTerm h_in = do
        ...
        return RunTerm {
            ...,
            termOps = Right FileOps {
                wrapFileInput = hWithBinaryMode h_in,
                withoutInputEcho = hWithoutInputEcho h_in,
               ...
              }
         }
      
where `hWithoutInputEcho` is the function from your `Echo` module.  That way the `FileOps` implicitly captures whichever `Handle` was specified in the `Backend`.

Finally, replace the code from `getPassword` https://github.com/judah/haskeline/blob/master/System/Console/Haskeline.hs#L246
with:

    getPassword x = promptedInput (...)
                    (\fops ->  withoutInputEcho fops False
                                  $ runMaybeT $ getLocaleLine fops)

The value of fops (and thus, the Handle that it wraps) that gets piped in will depend on which backend is used (the logic is in System.Console.Haskeline.Backend).  For reference (you don't need to change anything), these are the places that pass in the handle to it:
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend/Win32.hsc#L382
https://github.com/judah/haskeline/blob/master/System/Console/Haskeline/Backend.hs#L54